Documentation/Evolution/CharacterClasses.md

```diff
@@ -0,0 +1,354 @@
+- `\uhhhh`: Must use 4 hexidecimal digits.
+- `\x{hhh...}`: Can use 1 or more hexidecimal digits.
+- `\Uhhhhhhhh`: Must use 8 hexidecimal digits.
+
+**DSL:** This functionality is supported through the DSL's use of string literals.
+
+```swift
+Pattern {
+    "abc"
+    "\u{032}"
+    "def"
+}
+```
+
+### Unicode named values and properties: `\N`, `\p`, `\P`
+
+`\N{NAME}` matches a Unicode scalar value with the specified name. For example, `/\p{name=HANGUL SYLLABLE GAG}/` matches the character `"각"`.
+
+`\p{PROPERTY}` and `\p{PROPERTY=VALUE}` match a Unicode scalar value with the given Unicode property (and value, if given). Binary properties (such as `White_Space`) and either full or abbreviated general categories (e.g. `Uppercase_Letter` or `Lu`), can be used without a value. For other properties, provide the property and value to match. For example, `/\p{Numeric_Type=Decimal}+/` matches the full range of the string `"0① ₂³"`.
+
+While most Unicode-defined properties can only match at the Unicode scalar level, some are defined to match an extended grapheme cluster. For example, `/\p{RGI_Emoji_Flag_Sequence}/` will match any flag emoji character, which are composed of two Unicode scalar values.
+
+`\P{...}` matches the inverse of the characters matched by the corresponding `\p{...}` character class.
+
+**DSL:** This functionality is supported through APIs that take a predicate as a parameter, focused either on a `Character` or `UnicodeScalar` instance.
+
+```swift
+Pattern {
+    "abc"
+    OneOrMore(.unicodeProperty(where: \.isAlphabetic))
+    "def"
+}
+```
+
+### POSIX Character Classes: `[:NAME:]`
+
+The following POSIX character classes, of the form `[:NAME:]`, are supported:
+
+| POSIX class  | Matches           |
+|--------------|-------------------|
+| `[:alnum:]`  | `[A-Za-z0-9]`     |
+| `[:alpha:]`  | `[A-Za-z]`        |
+| `[:ascii:]`  | `[\x00-\x7F]`     |
+| `[:blank:]`  | `[ \t]`           |
+| `[:cntrl:]`  | `[\x00-\x1F\x7F]` |
+| `[:digit:]`  | `[0-9]`           |
+| `[:graph:]`  | `[\x21-\x7E]`     |
+| `[:lower:]`  | `[a-z]`           |
+| `[:print:]`  | `[[:graph:] ]`    |
+| `[:punct:]`  | `[-!"#$%&'()*+,./:;<=>?@[\\\]^_{|}~]` |
+| `[:space:]`  | `[ \t\n\r\f\v]`   |
+| `[:upper:]`  | `[A-Z]`           |
+| `[:word:]`   | `[A-Za-z0-9_]`    |
+| `[:xdigit:]` | `[0-9A-Fa-f]`     |
+
+### Custom Classes: `[...]`
+
+Users can create custom character classes by using literal characters and most predefined character classes within square brackets.
+
+- Individual characters and character classes described above can be used as is.
+- Ranges of Unicode scalar values can be specified by separating the start and end of the range (inclusive) with a hyphen. Note that because `Character` isn't a `Comparable` type, ranges can only match Unicode scalar values. Crossing the UTF-16 surrogate pair range (U+D800 to U+DFFF) is permitted, but values in that range do not match anything, regardless of the string's encoding.
```

Comment:
I'm pretty sure `Character` is `Comparable`:

```swift
1> Character("A") < Character("B")
$R0: Bool = true
```

Also I think in the Regex Literal pitch we showed at least pseudo code that was proposing representing custom classes as a closed range of characters?

Reply:
Clarifying enumeration of grapheme clusters should be a main aim of this discussion.